DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/state_machine/src/state_machine.hpp

```diff
@@ -27,57 +27,61 @@ using StreamingPipeProps = decltype(sycl::ext::oneapi::experimental::properties(
     sycl::ext::intel::experimental::protocol<
         sycl::ext::intel::experimental::protocol_name::
             avalon_streaming_uses_ready>));
+
 using MyStreamingBeat =
     sycl::ext::intel::experimental::StreamingBeat<float, true, false>;
 
 using StreamIn_NaiveStateMachine =
     sycl::ext::intel::experimental::pipe<IDStreamDataIn_NaiveStateMachine, MyStreamingBeat, 0,
                                          StreamingPipeProps>;
-using StreamIn_ProperStateMachine =
-    sycl::ext::intel::experimental::pipe<IDStreamDataIn_ProperStateMachine, MyStreamingBeat, 0,
+using StreamIn_OptimizedStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataIn_OptimizedStateMachine, MyStreamingBeat, 0,
                                          StreamingPipeProps>; 
 using StreamOut_NaiveStateMachine =
     sycl::ext::intel::experimental::pipe<IDStreamDataOut_NaiveStateMachine, MyStreamingBeat, 0,
                                          StreamingPipeProps>;
-using StreamOut_ProperStateMachine =
-    sycl::ext::intel::experimental::pipe<IDStreamDataOut_ProperStateMachine, MyStreamingBeat, 0,
+using StreamOut_OptimizedStateMachine =
+    sycl::ext::intel::experimental::pipe<IDStreamDataOut_OptimizedStateMachine, MyStreamingBeat, 0,
                                          StreamingPipeProps>;
 
 // function for task_sequence
-float updateCoeff(float *coeffPtr, int *executionCount) {
-  float coeffNew = coeffPtr[0];
-  executionCount[0] = executionCount[0] + 1;
-  return coeffNew;
+float UpdateCoeff(float *coeff_ptr, int *execution_count) {
+  float coeff_new = coeff_ptr[0];
+  // read-modify-write of external memory will result in a poor II.
+  execution_count[0] = execution_count[0] + 1;
+  return coeff_new;
 }
 
-float compute(float coeff, float data) {
+// function for task_sequence
+float Compute(float coeff, float data) {
   return coeff * data;
 }
 
```

Comment:
```suggestion
template<typename StreamIn, typename StreamOut>
float Process(float coeff) {
          MyStreamingBeat beat = StreamIn::read();
          // multiplication of input data with a coefficient can occur with II=1
          beat.data = beat.data * coeff;
          StreamOut::write(beat);
}
```
If we move *all* the processing from the state machine into the task function, then we can remove the `get()` call. Does this allow II=1 during simulation?